crates/syn2mas/src/mas_writer.rs

```diff
@@ -0,0 +1,656 @@
+    /// - If committing the changes failed.
+    ///
+    /// # Panics
+    ///
+    /// - If connections were not returned to the pool. (This indicates a serious bug.)
+    pub async fn finish(self) -> Result<(), Vec<Error>> {
+        let mut errors = Vec::new();
+
+        let Self {
+            num_connections,
+            mut connection_rx,
+            connection_tx,
+        } = self;
+        // Drop the sender handle so we gracefully allow the receiver to close
+        drop(connection_tx);
+
+        let mut finished_connections = 0;
+
+        while let Some(connection_or_error) = connection_rx.recv().await {
+            finished_connections += 1;
+
+            match connection_or_error {
+                Ok(mut connection) => {
+                    if let Err(err) = query("COMMIT;").execute(&mut connection).await {
+                        errors.push(err.into_database("commit writer transaction"));
+                    }
+                }
+                Err(error) => {
+                    errors.push(error);
+                }
+            }
+        }
+        assert_eq!(finished_connections, num_connections, "syn2mas had a bug: connections went missing {finished_connections} != {num_connections}");
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    }
+}
+
+pub struct MasWriter<'c> {
+    conn: LockedMasDatabase<'c>,
+    writer_pool: WriterConnectionPool,
+
+    indices_to_restore: Vec<IndexDescription>,
+    constraints_to_restore: Vec<ConstraintDescription>,
+}
+
+pub struct MasNewUser {
+    pub user_id: Uuid,
+    pub username: String,
+    pub created_at: DateTime<Utc>,
+    pub locked_at: Option<DateTime<Utc>>,
+    pub can_request_admin: bool,
+}
+
+pub struct MasNewUserPassword {
+    pub user_password_id: Uuid,
+    pub user_id: Uuid,
+    pub hashed_password: String,
+    pub created_at: DateTime<Utc>,
+}
+
+/// List of all MAS tables that are written to by syn2mas.
+pub const MAS_TABLES_AFFECTED_BY_MIGRATION: &[&str] = &["users", "user_passwords"];
+
+/// Detect whether a syn2mas migration has started on the given database.
+///
+/// Concretly, this checks for the presence of syn2mas restoration tables.
+///
+/// Returns `true` if syn2mas has started, or `false` if it hasn't.
+///
+/// # Errors
+///
+/// Errors are returned under the following circumstances:
+///
+/// - If any database error occurs whilst querying the database.
+/// - If some, but not all, syn2mas restoration tables are present.
+///   (This shouldn't be possible without syn2mas having been sabotaged!)
+pub async fn is_syn2mas_in_progress(conn: &mut PgConnection) -> Result<bool, Error> {
+    // Check if there is a resumption table...
+    let num_resumption_tables = query!(
+        r#"
+        SELECT 1 AS _dummy FROM pg_tables WHERE schemaname = current_schema
+        AND tablename IN ('syn2mas_restore_constraints', 'syn2mas_restore_indices')
```

Comment:
Maybe define an array of tables, pass it as query parameter so that you can use `len` after?